gsy_framework/matching_algorithms/pay_as_bid_matching_algorithm.py

```diff
@@ -15,33 +15,78 @@ class PayAsBidMatchingAlgorithm(BaseMatchingAlgorithm):
            since the most affordable offers will be allocated for the most aggressive buyers.
     """
 
+    @staticmethod
+    def _match_one_bid_one_offer(offer: Dict, bid: Dict, already_selected_order_energy: Dict,
+                                 market_id: str, time_slot: str) -> Optional[BidOfferMatch]:
+        """
+        Try to match one bid with one offer, and at the same time update the dict with the
+        already selected order energy in order to be able to reuse the same order in future
+        matches.
+        """
+        if (offer.get("energy_rate") - bid.get("energy_rate")) > FLOATING_POINT_TOLERANCE:
+            return None
+
+        if bid["id"] not in already_selected_order_energy:
+            already_selected_order_energy[bid["id"]] = bid["energy"]
+        if offer["id"] not in already_selected_order_energy:
+            already_selected_order_energy[offer["id"]] = offer["energy"]
+        offer_energy = already_selected_order_energy[offer["id"]]
+        bid_energy = already_selected_order_energy[bid["id"]]
```

Comment:
Nit.
```suggestion
        offer_energy = offer["energy"]
        bid_energy = bid["energy"]
        if bid["id"] not in already_selected_order_energy:
            already_selected_order_energy[bid["id"]] = bid_energy
        if offer["id"] not in already_selected_order_energy:
            already_selected_order_energy[offer["id"]] = offer_energy

```

Reply:
These are not equivalent actually. `offer["energy"]` and `bid["energy"]` are the energy from the original offer and bid respectively. However, we might have already included this energy in other matches, therefore we really need to use the already_selected_order_energy values (if they are there of course). If the dict does not contain the orders yet, then we anyways need to populate it, that's why we add the corresponding entries to the dict.

Reply:
🤦  Indeed, I forgot that the `already_selected_order_energy` is a refference to an instance outside of this function and it might be updated there. Thanks for explaining!